### PR TITLE
Add `--query-range.request-downsampled` flag to Query Frontend (#2641)

### PR DESCRIPTION
**Pull request by krya-kryak**
Signed-off-by: Vladimir Kononov <krya-kryak@users.noreply.github.com>

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

- [#2641](https://github.com/thanos-io/thanos/issues/2641) Query Frontend: Added `--query-range.request-downsampled` flag enabling additional queries for downsampled data in case of empty or incomplete response to range request.

This is needed to zoom into past, where raw and/or 5m datapoints could be lacking due to retention policies.

Q: Why query-frontend instead of query?
A: Query currently has no concept of modifying queries, while query-frontend does. Secondly, at the moment query seems to have no cortex dependencies, and cortex's matrix merger is a convenient thing.

## Verification

Here are screenshots for the same queries made against the very same underlying thanos-query. Please excuse the quality. For those struggling to make sense of the pixelated mess, all requests are the same, all set to a particular time window in the past, the only difference is downsample picker, set to "Max 5m downsampling", "Max 1h downsampling", "auto downsampling" respectively. 5m points are half-gone at this point due to retention strategy, while 1h points are still present. Note "auto downsampling" actually producing sensible graph preserving maximum amount of infromation: 
before-after

### CODE REVIEW

**Comment by krya-kryak** [comment]
sidenote: to make use of this in grafana, one currently needs to add `max_source_resolution=auto` to custom query parameters.

**Review by bwplotka — APPROVED** [review]
I love it. Amazing work!

Simple and effective. 

However this fix feels like workaround for broken Querier logic (: I think it's fine to merge it here - I think it's useful, but wonder if we can have similar logic as enhanced Querier -> Store Gateway better handling or something.

I just want to make sure we fix the root cause of this bad user experience 🤗 Let's discuss this. Reason, why it's so imperfect, is that we did not have time to look into this part for a long time. Help wanted (:

File: pkg/queryfrontend/downsampled.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+		}
+	})
+}
+
+type downsampled struct {
+	next   queryrange.Handler
+	merger queryrange.Merger
+
+	// Metrics.
+	extraCounter prometheus.Counter
```
[the remark this answers is not in the document]
Comment (krya-kryak) [reply pkg/queryfrontend/downsampled.go]: Actually, it wasnt' a reference to type, just poor naming on my part. Thanks for the suggestion :)

**Comment by bwplotka** [comment]
Reran tests

**Pull request merged**